<h1 align="center">Python performance exercises</h1>

## Python best practices exercises

### Exercise 1

considering the following function for concatenating list strings with delimiter.

In [1]:
def ft_concatenate(l_strings, d):
    """concatenate list of strings into one string separated by delimiter"""
    res = l_strings[0]
    for e in l_strings[1:]:
        res = res + d + e
    return res

- profile the function and identify the bottlenecks.
- improve speed up of the function
*Hint: you may need to look to the string functions in python documentation*

In [3]:
# Utils
import random
import string

def generate_string_list(n, length):
    string_list = []
    for i in range(n):
        characters = string.ascii_letters + string.digits + " "
        ramdom_string = ''.join(random.choice(characters) for _ in range(length))
        string_list.append(ramdom_string)
    return string_list

In [4]:
# data
text = generate_string_list(50000, 10)
delimiter = " * "

In [6]:
# Let profil with timeit
%timeit ft_concatenate("text"*1000, delimiter)

1.13 ms ± 10.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [7]:
# Let profil with prun
%prun -s cumulative ft_concatenate("text"*1000, delimiter)

         4 function calls in 0.001 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.001    0.001 {built-in method builtins.exec}
        1    0.000    0.000    0.001    0.001 <string>:1(<module>)
        1    0.001    0.001    0.001    0.001 2524129870.py:1(ft_concatenate)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}

In [8]:
!pip install line_profiler
%load_ext line_profiler

In [9]:
%lprun -f ft_concatenate ft_concatenate("text"*1000, delimiter)

Timer unit: 1e-09 s

Total time: 0.004327 s
File: /var/folders/w8/b_cflrn97k9c15rcrn5t8mmr0000gn/T/ipykernel_7348/2524129870.py
Function: ft_concatenate at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           def ft_concatenate(l_strings, d):
     2                                               """concatenate list of strings into one string separated by delimiter"""
     3         1       2000.0   2000.0      0.0      res = l_strings[0]
     4      4000     679000.0    169.8     15.7      for e in l_strings[1:]:
     5      3999    3645000.0    911.5     84.2          res = res + d + e
     6         1       1000.0   1000.0      0.0      return res

In [11]:
!pip install memory_profiler
%load_ext memory_profiler

The memory_profiler extension is already loaded. To reload it, use:
  %reload_ext memory_profiler


In [15]:
%mprun -f ft_concatenate ft_concatenate("text"*1000, delimiter)

UsageError: Could not find function 'ft_concatenate("text"*1000,'.
SyntaxError: '(' was never closed (<string>, line 1)


In [16]:
# What we have to do is to minimize the number of string concatenations: Here is an optimize version of the function
def ft_concatenate(l_strings, d):
    """concatenate list of strings into one string separated by delimiter"""
    return d.join(l_strings)

In [17]:
# Proof
%lprun -f ft_concatenate ft_concatenate(text, delimiter)

Timer unit: 1e-09 s

Total time: 0.002093 s
File: /var/folders/w8/b_cflrn97k9c15rcrn5t8mmr0000gn/T/ipykernel_7348/4137348578.py
Function: ft_concatenate at line 2

Line #      Hits         Time  Per Hit   % Time  Line Contents
     2                                           def ft_concatenate(l_strings, d):
     3                                               """concatenate list of strings into one string separated by delimiter"""
     4         1    2093000.0    2e+06    100.0      return d.join(l_strings)

<div style="padding: .5em; background: #374151; color: #fff; width: 100%">
<h5 style="color:red">Conclusion:</h5>
With this version we could see rthat with just one hit and time=783000. we get the result. And this is much better that oldier result.
</div>

### Exercise 2

In this exercise you will solve the following problem using two methods bruteforce method, and fast method.

**Problem:** You are given a list of n integers, and your task is to calculate the number of distinct values in the list.

**Example**
- Input:
5
2 3 2 2 3

- Output:
2

**Implement the following methods:**

1. **bruteforce method:** create an empty list and start adding items for the given list without adding the previous item add, at the end the result list will contain unique values, print lenght of the list and you are done. 
2. **fast method** think of using Set data structure.

- time the two methods, what do you think?

In [18]:
# bruteforce method
def bruteforce(l):
    distinct_elements = []
    for element in l[1:]:
        if element not in distinct_elements:
            distinct_elements.append(element)
    return len(distinct_elements)

bruteforce([5, 2, 3, 2, 2, 3])


2

In [19]:
# fast method
def fast(l):
    distinct_elements = set(l[1:])
    return len(distinct_elements)

fast([5, 2, 3, 2, 2, 3])

2

In [20]:
import time

In [23]:
# Create a random list of numbers for testing
random_integers = [random.randint(0, 9) for _ in range(10)]
random_integers.insert(0, 10)

# time the two methods
start_time = time.time()
bruteforce(random_integers)
end_time = time.time()
elapsed_time1 = (end_time - start_time)

start_time = time.time()
fast(random_integers)
end_time = time.time()
elapsed_time2 = end_time - start_time

print(f"Bruteforce: \t{elapsed_time1:.6f} seconds\nFast: \t\t{elapsed_time2:.6f} seconds")

Bruteforce: 	0.000030 seconds
Fast: 		0.000027 seconds


## Cython exercises

### Exercise 1

1. load the cython extension.

In [24]:
!pip install cython
%load_ext cython

2. Considering the following polynomial function:

In [25]:
def poly(a,b):
    return 10.5 * a + 3 * (b**2)

- Create an equivalent Cython function of `poly` with name `poly_cy`.

In [27]:
%%cython -a
cpdef double poly_cy(double a, double b):
    return 10.5 * a + 3 * (b**2)

3. time the performance of Python and Cython version of the function, what is the factor of speed up between the two verions.

In [28]:
# write your code here
%timeit poly(34,456)
%timeit poly_cy(34,456)


255 ns ± 5.45 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
55.9 ns ± 0.552 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [31]:
354/80.6

4.392059553349876

<div style="padding: .5em; background: #374151; color: #fff; width: 100%">
<h5 style="color:red">Conclusion:</h5>
The factor of speed up is = 4.39
</div>

4. Now let's work on another example using loop.
    - rewrite the same function below fib that calculates the fibonacci sequence using cython, but now try to add type for the variables used inside it, add a prefix `_cy` to your new cython function.

In [29]:
def fib(n):
    a, b = 1, 1
    for i in range(n):
        a, b = a + b, a

    return a

In [34]:
%%cython -a
cpdef int fib_cy(int n):
    cdef int a, b, i
    a, b = 1, 1
    for i in range(n):
        a, b = a + b, a
    return a

- time the two function for fibonacci series, with n = 20, what is the factor of speed now, What do you think?

In [31]:
%timeit fib(20)
%timeit fib_cy(20)

933 ns ± 39.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
44.9 ns ± 0.299 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [35]:
factor = 1.24*1000 / 61.1
factor

20.29459901800327

5. Recursive functions are functions that call themselves during their execution. Another interesting property of the Fibonacci sequence is that it can be written as a recursive function. That’s because each item depends on the values of other items (namely item n-1 and item n-2)

- Rewrite the fib function using recursion. Is it faster than the non-recursive version? Does Cythonizing it give even more of an advantage? 

In [35]:
def fib_recursive(n):
    if n <= 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fib_recursive(n - 1) + fib_recursive(n - 2)


In [36]:
%%cython -a
cpdef int fib_recursive_cy(int n):
    if n <= 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fib_recursive_cy(n - 1) + fib_recursive_cy(n - 2)

In [44]:
from numba import njit, prange

@njit(fastmath=True, parallel=True)
def fib_numba(n: "int"):
    a, b = 1, 1
    for i in prange(n):
        a, b = a + b, a

    return a

In [43]:
%timeit fib(20)
%timeit fib_cy(20)
%timeit fib_recursive(20)
%timeit fib_recursive_cy(20)
%timeit fib_numba(20)

925 ns ± 7.42 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
45.1 ns ± 0.29 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
1.87 ms ± 6.57 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
47.7 µs ± 262 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
13.4 µs ± 6.67 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [45]:
%timeit fib_numba(20)

13.2 µs ± 480 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


### Exercise 2

- Monte Carlo methods are a broad class of computational algorithms that rely on repeated random sampling to obtain numerical results. 
- One of the basic examples of getting started with the Monte Carlo algorithm is the estimation of Pi.

**Estimation of Pi**

- The idea is to simulate random (x, y) points in a 2-D plane with domain as a square of side 1 unit. 
- Imagine a circle inside the same domain with same diameter and inscribed into the square. 
- We then calculate the ratio of number points that lied inside the circle and total number of generated points. 
- Refer to the image below:

![demo](../data/MonteCarloPlot.png)

We know that area of the square is 1 unit sq while that of circle is $\pi \ast  (\frac{1}{2})^{2} = \frac{\pi}{4}$. Now for a very large number of generated points,

![demo](../data/MonteCarloCalc.png)


## The Algorithm

1. Initialize cile_points, square_points and interval to 0.
2. Generate random point x.
3. Generate random point y.
4. Calculate d = x*x + y*y.
5. If d <= 1, increment circle_points.
6. Increment square_points.
7. Increment interval.
8. If increment < NO_OF_ITERATIONS, repeat from 2.
9. Calculate pi = 4*(circle_points/square_points).
10. Terminate.

**Your mission:** time the function `monte_carlo_pi`, identify the bottlenecks and create a new version using cython functionality to speed up monte carlo simulation for PI, use 100,000 points and compare the speed up factor between python and cython, considering the following optimizations:
- add type for variables used.
- add type for the function
- use c rand function instead of python rand function.
 
*Hint: you can import function from C libraries using the following approach `from libc.<name of c library> cimport <library function name>`, replace the holders `<>` with the right identities for the current problem*

In [47]:
import random
import numpy as np

def monte_carlo_pi(nsamples):
    pi = 0.
    circle_points = 0
    square_points = 0
    interval = 0

    while (interval < nsamples):
        x, y = random.random(), random.random()
        d = np.sqrt(x**2 + y**2)

        if d <= 1:
            circle_points += 1
        square_points += 1
        interval += 1

    pi = 4 * (circle_points / square_points)
    
    return pi

In [52]:
monte_carlo_pi(100000)

3.13696

In [40]:
%timeit monte_carlo_pi(100000)

31.3 ms ± 352 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


Profiling

In [41]:
%prun -s cumulative monte_carlo_pi(100000)

         200004 function calls in 0.049 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.049    0.049 {built-in method builtins.exec}
        1    0.000    0.000    0.049    0.049 <string>:1(<module>)
        1    0.042    0.042    0.049    0.049 4252419539.py:2(monte_carlo_pi)
   200000    0.007    0.000    0.007    0.000 {method 'random' of '_random.Random' objects}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}

In [42]:
%reload_ext line_profiler
%lprun -f monte_carlo_pi monte_carlo_pi(100000)

Timer unit: 1e-09 s

Total time: 0.146672 s
File: /var/folders/w8/b_cflrn97k9c15rcrn5t8mmr0000gn/T/ipykernel_2672/4252419539.py
Function: monte_carlo_pi at line 2

Line #      Hits         Time  Per Hit   % Time  Line Contents
     2                                           def monte_carlo_pi(nsamples):
     3         1       1000.0   1000.0      0.0      pi = 0.
     4         1          0.0      0.0      0.0      circle_points = 0
     5         1          0.0      0.0      0.0      square_points = 0
     6         1          0.0      0.0      0.0      interval = 0
     7                                           
     8    100001   21619000.0    216.2     14.7      while (interval < nsamples):
     9    100000   28498000.0    285.0     19.4          x, y = random.random(), random.random()
    10    100000   29335000.0    293.4     20.0          d = x**2 + y**2
    11                                           
    12    100000   17664000.0    176.6     12.0          if d <= 1:
    1

In [43]:
%reload_ext memory_profiler
%mprun -f monte_carlo_pi monte_carlo_pi(100000)

ERROR: Could not find file /var/folders/w8/b_cflrn97k9c15rcrn5t8mmr0000gn/T/ipykernel_2672/4252419539.py



Optimized version

In [54]:
%%cython -a
from libc.stdlib cimport rand, RAND_MAX
cpdef double monte_carlo_pi_cy(int nsamples):
    cdef double pi, x, y, d
    pi = 0.
    cdef int circle_points, square_points, interval
    circle_points, square_points, interval = 0, 0, 0

    while (interval < nsamples):
        x, y = rand() / RAND_MAX, rand() / RAND_MAX
        d = (x**2 + y**2)**0.5

        if d <= 1:
            circle_points += 1
        square_points += 1
        interval += 1

    pi = 4 * (circle_points / square_points)
    
    return pi

Content of stderr:
/Users/genereux/.ipython/cython/_cython_magic_75499905a3509442f52bcf7499a65715acfef017.c:2327:7: warning: code will never be executed [-Wunreachable-code]
      PyErr_SetString(PyExc_ZeroDivisionError, "float division");
      ^~~~~~~~~~~~~~~
/Users/genereux/.ipython/cython/_cython_magic_75499905a3509442f52bcf7499a65715acfef017.c:2321:7: warning: code will never be executed [-Wunreachable-code]
      PyErr_SetString(PyExc_ZeroDivisionError, "float division");
      ^~~~~~~~~~~~~~~
2 warnings generated.

Comparaison

In [55]:
%timeit monte_carlo_pi(100000)
%timeit monte_carlo_pi_cy(100000)

96 ms ± 4.01 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
1.28 ms ± 14.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


## Numba exercises

### Exercise 1

Previously we considered how to approximateby Monte Carlo.

- Use the same idea here, but make the code efficient using Numba.
- Compare speed with and without Numba when the sample size is large.

In [56]:
from numba import njit

@njit
def monte_carlo_pi_numba(nsamples):
    pi = 0.
    circle_points = 0
    square_points = 0
    interval = 0

    while (interval < nsamples):
        x, y = random.uniform(-1, 1), random.uniform(-1, 1)
        d = (x**2 + y**2)**0.5

        if d <= 1:
            circle_points += 1
        square_points += 1
        interval += 1

    pi = 4 * (circle_points / square_points)
    
    return pi

Comparaison

In [57]:
%timeit monte_carlo_pi(100000)
%timeit monte_carlo_pi_cy(100000)
%timeit monte_carlo_pi_numba(100000)

94 ms ± 1.18 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
1.27 ms ± 10.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
734 µs ± 4.37 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


### Exercise 2

In the [Introduction to Quantitative Economics](https://python.quantecon.org/intro.html) with Python lecture series you can learn all about finite-state Markov chains.

For now, let's just concentrate on simulating a very simple example of such a chain.

Suppose that the volatility of returns on an asset can be in one of two regimes — high or low.

The transition probabilities across states are as follows ![markov](../data/markov.png)

For example, let the period length be one day, and suppose the current state is high.

We see from the graph that the state tomorrow will be

- high with probability 0.8

- low with probability 0.2

Your task is to simulate a sequence of daily volatility states according to this rule.

Set the length of the sequence to `n = 1_000_000` and start in the high state.

Implement a pure Python version and a Numba version, and compare speeds.

To test your code, evaluate the fraction of time that the chain spends in the low state.

If your code is correct, it should be about 2/3.

Hints:

- Represent the low state as 0 and the high state as 1.

- If you want to store integers in a NumPy array and then apply JIT compilation, use `x = np.empty(n, dtype=np.int_)`.
